### ISSUE TITLE
Strange values for BecaSmart Devices

### ISSUE THREAD

**Issue by baresel** [Apollon77/ioBroker.tuya#456]
I added a BecaSmart floor heating device to my smarthome and followed the instructions to setup the connector. The device and schema appears in the objects, but the temperature values are strange:

The tempSet (20°C) is shown in the admin as 4 â„ƒ, the tempCurrent (sensor shows 21,5 °C) is shown as 4.3 â„ƒ

Attached you will find the object tree for the sensor.

[tuya.0.bfc78f23ce7796482dos9e.json.txt](https://github.com/Apollon77/ioBroker.tuya/files/9695118/tuya.0.bfc78f23ce7796482dos9e.json.txt)

**Comment by Apollon77** [comment]
Hm ... in fact the values are exactly provided that way as it seems ... there is alo nothing else defined in the schema that should help ... when thinking more and calculating with the values you provided it seems that you need to multiply the value by 5 ... interstingly the 5 is set as "step" which is also bullshit ... in the end very strange.

Maybe this helps yu. I would add an alias with a calculation function that multiplies the value with 5 ... problem might be that because of the "step=5" it could be hard to set a value because "step" normally has a completely different meaning then it is used in this shema.

I try to get more details. so I leave this open, but I assue iI can not really do more.

**Comment by baresel** [comment]
Yes, the values provided from the device are a fifth of the "real" values. I can deal with it, but i have seen the adapter is showing a warning message each time it gets the values stating the read value of 4 is lower than the minimum of 10 ...
Where can i change this behavior to use other thresholds or use the multiplier ?

**Comment by Apollon77** [comment]
In fact or min/max/step I would simply edit the affected objects and remove these properties in common ... The adapter "trust" the data coming in the schema ... when they are wrong it is problematic. The multiplier can not be used directly .. you need a small javascript or you can create an alias of the object and use the read calculation function

**Comment by ikmd** [comment]
Same here, just installed a MOES wifi thermostat. All the non-boolean values are a bit messy.
The temperatures are as above 1/5 of the correct value, names and units are cryptic symbols ("â„ƒ" probably should be "°C"?).
Firmware versions are
Main: V2.0.2
MCU-Module: V4.0.1

**Comment by ikmd** [comment]
So, I just had a look at the code and though I don't know what I'm doing I did some changes locally on version 3.6.15.
First I modified lib/mapper.js and updated line 114 though I must say I don't know why this schema is chosen :) (mainly replaced "â„ƒ" with "°C").
Next I applied some multiplication/division to the values on given keys (as I'm currently only using one device, I'm not checking for device ID)
226:
```
            onChange = (value) => {
                if (id == 2) {
                    value = value / 5;
                }
```

556:
```
                    let value = data.dps[id];
                    if (id == 2 || id == 3) {
                        value = value * 5;
                    }
```
Seems to be a temporary fix for me.

**Comment by Apollon77** [comment]
1.) unit: The units are used as coming from Schema. The adapter can not automatically translate them. SO you need to semantically interpret them

2.) Yes this change might work for just your special case,but in fact it nothing in general and will be re-set with each adater update again. So not that helpfull for you. An Alias that recalculated the value is much more simple and futureproof but your decision

**Comment by ikmd** [comment]
Ah thanks, I only just started using iobroker don't know anything about aliases yet. Will have a look!

**Comment by morskipas** [comment]
> So, I just had a look at the code and though I don't know what I'm doing I did some changes locally on version 3.6.15. First I modified lib/mapper.js and updated line 114 though I must say I don't know why this schema is chosen :) (mainly replaced "â„ƒ" with "°C"). Next I applied some multiplication/division to the values on given keys (as I'm currently only using one device, I'm not checking for device ID) 226:
> 
> ```
>             onChange = (value) => {
>                 if (id == 2) {
>                     value = value / 5;
>                 }
> ```
> 
> 556:
> 
> ```
>                     let value = data.dps[id];
>                     if (id == 2 || id == 3) {
>                         value = value * 5;
>                     }
> ```
> 
> Seems to be a temporary fix for me.

Hi ikmd,
short question, can you pls help.
I have exactly the same error and wanted to try your workaround. My file mapper.js in the directory /opt/iobroker/node_modules/iobroker.tuya/lib does not provide the possibility to replace the unit at all. I don't have a line 226 and 556 either. Did you make all the changes in the mapper.js file?

Thank you!
Frank

**Comment by ikmd** [comment]
Hi @morskipas, no this was done one level up in main.js.
But as Apollon pointed out, it can make more sense to set up the aliases for the affected devices.
I used the "devices" adapter to create an alias. After installing, you get a new tab where you can set up a new thermostat alias. There you need to define at least "set" and "actual" values, clicking on "fx" allows you to multiply by 5. Let me know if you need more help.
![Screenshot_2022-10-09-15-38-19-02_4641ebc0df1485bf6b47ebd018b5ee76](https://user-images.githubusercontent.com/6737643/194760559-3f0925a1-a61a-47a9-aa5d-c56cb55e9c10.jpg)
![Screenshot_2022-10-09-15-39-15-35_4641ebc0df1485bf6b47ebd018b5ee76](https://user-images.githubusercontent.com/6737643/194760567-f78bf058-4075-4a9b-b2a2-c4474cd1ee97.jpg)

@Apollon77 FYI: since I changed mapper.js and re-added the device, the value of the 101 (program) field changed from something like "(((((,(((" (and a lot of unreadable characters) to Latin characters looking like "AAYoAAgeHgseHg0eABEsA..." (this is trunkated, not the actual value).

**Comment by morskipas** [comment]
> Hi @morskipas, no this was done one level up in main.js. But as Apollon pointed out, it can make more sense to set up the aliases for the affected devices. I used the "devices" adapter to create an alias. After installing, you get a new tab where you can set up a new thermostat alias. There you need to define at least "set" and "actual" values, clicking on "fx" allows you to multiply by 5. Let me know if you need more help. ![Screenshot_2022-10-09-15-38-19-02_4641ebc0df1485bf6b47ebd018b5ee76](https://user-images.githubusercontent.com/6737643/194760559-3f0925a1-a61a-47a9-aa5d-c56cb55e9c10.jpg) ![Screenshot_2022-10-09-15-39-15-35_4641ebc0df1485bf6b47ebd018b5ee76](https://user-images.githubusercontent.com/6737643/194760567-f78bf058-4075-4a9b-b2a2-c4474cd1ee97.jpg)
> 
> @Apollon77 FYI: since I changed mapper.js and re-added the device, the value of the 101 (program) field changed from something like "(((((,(((" (and a lot of unreadable characters) to Latin characters looking like "AAYoAAgeHgseHg0eABEsA..." (this is trunkated, not the actual value).

Hi ikmd,
thx a lot for the quick answer :-)
I did it as described. What should be the result now? The values in the Tuya adapter is the same for me, I had already adjusted the unit degrees Celsius. Under objects/devices I see only admin, but no devices. I'm not sure if the devices adapter works correctly, I can't delete devices. Should the result be the following? I have under objects/devices a datapoint which I can then address correctly, e.g. with 25 instead of 5?

Thanks for your help :-)
Frank
<img width="1667" alt="001" src="https://user-images.githubusercontent.com/6042276/194765255-9ab67424-00e4-4626-bd78-756fb1f717de.png">
<img width="1204" alt="002" src="https://user-images.githubusercontent.com/6042276/194765257-6222daa6-0375-45b1-9b98-fb743de33d8c.png">
<img width="1542" alt="003" src="https://user-images.githubusercontent.com/6042276/194765259-422950dd-5891-428a-a502-9bbf53698d25.png">
<img width="1534" alt="004" src="https://user-images.githubusercontent.com/6042276/194765260-e5a2678d-fb68-4bda-a513-64c1f6653bee.png">

**Comment by morskipas** [comment]
> @morskipas Looks good. Thoses aliases you created are under alias.0.Tuya. You can see the folder structure in your first screenshot underneath Tuya and Dummy. Changing the value of alias.0.Tuya.Dummy.SET to 25 should update the corresponding Tuya device to 5 now.

@ikmd Hi ikmd,

ohh my goodness, I always searched for devices/0/tuya/... in the object tree. but it is now displayed under alias/0/tuya/...  
Now I know, thanks for your help and this on Sunday.

Enjoy the evening!
Frank
